[DONE] Constraints configurable on per-weight basis, richer API

maxpumperla opened the pull request:
## What changes were proposed in this pull request?

Contraints can now be set on weights only, biases only or both. On request, one can also constrain custom parameters by name. For instance for LSTMs, we can do
```
                    .layer(new LSTM.Builder().nIn(12).nOut(10)
                            .recurrentConstraints(lc).build())
```
to constrain "RW", i.e. recurrent weights. Added more tests for all of this.

Note that concerns of "what constraint to use" and "where to apply it to" have been separated.

AlexDBlack commented on deeplearning4j-core/src/test/java/org/deeplearning4j/gradientcheck/CNN1DGradientCheckTest.java:
```diff
@@ -51,7 +51,7 @@ public void testCnn1DWithZeroPadding1D() {
         int zeroPadding = 2;
         int paddedLength = length + 2 * zeroPadding;
 
-        Activation[] activations = {Activation.SIGMOID, Activation.TANH};
```
I want this one to keep failing so it doesn't get forgotten/missed :)

maxpumperla replied:
Since we have an open ticket with you assigned already (https://github.com/deeplearning4j/deeplearning4j/issues/3975) and I'd rather keep the build green, can we agree to leave it like this?

AlexDBlack commented on deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/api/layers/LayerConstraint.java:
```diff
@@ -4,11 +4,24 @@
+     * @param weightConstraint if contraint is applied to weight parameters
+     * @param paramNames Parameter names to which to apply the constraints
+     */
+    void applyConstraint(Layer layer, int iteration, int epoch, Boolean biasConstraint,
```
I'd use ```boolean``` instead of ```Boolean``` - shouldn't allow null here on the public API IMO.

AlexDBlack commented on deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/layers/Layer.java:
```diff
@@ -78,12 +79,18 @@
     protected List<LayerConstraint> constraints;
+    protected Boolean hasBiasConstraints;
+    protected Boolean hasWeightConstraints;
+    protected Set<String> constraintParamNames;
```
I don't like this. The parameter names that the constraint applies to should be in the constraints themselves, not on the layer. More generally we should have only *one* field related to constraints, with all relevant information contained within it.

AlexDBlack commented on deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/layers/Layer.java:
```diff
@@ -278,8 +301,29 @@ public T constraints(LayerConstraint... constraints) {
+         * @return Constraints
+         */
+        public T constraints(List<LayerConstraint> constraints, boolean hasBiasConstraints,
+                             boolean hasWeightConstraints, Set<String> constraintParamNames) {
```
This approach won't work: we can't use ```weightConstraints(x).biasConstraints(y)``` as the latter will override the former (i.e., x will be removed entirely, and only y will remain).
And again, the weight vs. bias vs. recurrent weight selection should IMO be in the constraints themselves, not as fields on the layer.

maxpumperla commented:
@AlexDBlack I think I addressed your concerns. We can now create very customised constraints, but to normal users it looks like a constraint is not actually linked to parameters (which internally they are). So it feels a little like keras, while we get around having to store different sets of constraints internally.

AlexDBlack commented on deeplearning4j-core/src/test/java/org/deeplearning4j/nn/conf/constraints/TestConstraints.java:
```diff
@@ -21,21 +22,401 @@
+        };
+
+        Set<String> customConstraint = new HashSet<>();
+        customConstraint.add("RW");
```
This customConstraint set isn't used anywhere?
We probably want' to check that the constraints are *not* applied to the other parameters too.

AlexDBlack commented on deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/constraint/BaseConstraint.java:
```diff
@@ -20,18 +22,18 @@
 
     protected boolean applyToWeights;
     protected boolean applyToBiases;
+    protected Set<String> paramNames = new HashSet<>();
```
It should be IMO one or the other - applyToWeights/applyToBiases OR paramNames set - not both. My vote is for the latter only.

AlexDBlack commented on deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/layers/LSTM.java:
```diff
@@ -83,6 +85,20 @@ public LayerMemoryReport getMemoryReport(InputType inputType) {
     public static class Builder extends AbstractLSTM.Builder<Builder> {
 
+
+        public Builder recurrentConstraints(LayerConstraint... constraints) {
```
I was thinking, and I'm not 100% sure on this, but maybe we name everything ```constraintsX``` where X is {Weights, Recurrent, Bias} etc. That way all constraint methods show up together in the javadoc and code suggestions in IDEs.

AlexDBlack commented on deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/NeuralNetConfiguration.java:
```diff
@@ -1245,8 +1246,14 @@ public Builder convolutionMode(ConvolutionMode convolutionMode) {
+                constraint.setApplyToWeights(true);
+                currentConstraints.add(constraint);
+            }
+            return constraints(currentConstraints);         }
```
I'm not sure defaulting  to biases on the .constraints(...) method is a good idea?
I'd probably go for .constraints(LayerContstraint...) method requiring the user to manually specify the weight/bias keys (i.e., the weightConstraints and biasConstraint methods become the 'default' ones to use).

This current approach will also override whatever the user has set...

maxpumperla replied:
@AlexDBlack what do you mean by "what the user has set"? This IS what the user sets. You mean what he sets on individual layers? There has to be some rule for what takes precedence and I think it makes sense that global config overrides local.

maxpumperla commented:
@AlexDBlack this is done from my side. All keras constraints are supported in model import. The nitpick you have above regarding constraint fields is understandable, but I don't feel like it justifies pouring more time into this right now.
